test/test.js

```diff
@@ -0,0 +1,358 @@
+				if (!(plaintext && to_hex(plaintext) == v.m)) throw new Error('Unexpected plaintext');
+			}
+		};
+
+		t.hash = function(){
+			var sha512Vectors = [
+				{
+					m: 'testing\n',
+					h: '24f950aac7b9ea9b3cb728228a0c82b67c39e96b4b344798870d5daee93e3ae5931baae8c7cacfea4b629452c38026a81d138bc7aad1af3ef7bfd5ec646d6c28'
+				},
+				{
+					m: 'The Conscience of a Hacker is a small essay written January 8, 1986 by a computer security hacker who went by the handle of The Mentor, who belonged to the 2nd generation of Legion of Doom.',
+					h: 'a77abe1ccf8f5497e228fbc0acd73a521ededb21b89726684a6ebbc3baa32361aca5a244daa84f24bf19c68baf78e6907625a659b15479eb7bd426fc62aafa73'
+				}
+			];
+			var sha256Vectors = [
+				{
+					m: 'testing\n',
+					h: '12a61f4e173fb3a11c05d6471f74728f76231b4a5fcd9667cef3af87a3ae4dc2'
+				},
+				{
+					m: 'The Conscience of a Hacker is a small essay written January 8, 1986 by a computer security hacker who went by the handle of The Mentor, who belonged to the 2nd generation of Legion of Doom.',
+					h: '71cc8123fef8c236e451d3c3ddf1adae9aa6cd9521e7041769d737024900a03a'
+				}
+			];
+
+			if (sodium.crypto_hash) for (var i = 0; i < sha512Vectors.length; i++) sha512_test(sha512Vectors[i]);
+			if (sodium.crypto_hash_sha256) for (var i = 0; i < sha256Vectors.length; i++) sha256_test(sha256Vectors[i]);
+
+			function sha512_test(v){
+				var hash = sodium.crypto_hash(v.m);
+				if (to_hex(hash) != v.h) throw new Error('Unexpected hash value for vector ' + JSON.stringify(v));
+			}
+
+			function sha256_test(v){
+				var hash = sodium.crypto_hash_sha256(v.m);
+				if (to_hex(hash) != v.h) throw new Error('Unexpected hash value for vector ' + JSON.stringify(v));
+			}
+
+		};
+
+		t.hmac = function(){
+			var hmacSHA512256Vectors = [
+				{
+					k: 'Jefe',
+					m: 'what do ya want for nothing?',
+					h: '164b7a7bfcf819e2e395fbe73b56e0a387bd64222e831fd610270cd7ea250554'
+				}
+			];
+
+			var hmacSHA512Vectors = [
+				{
+					k: 'Jefe',
+					m: 'what do ya want for nothing?',
+					h: '164b7a7bfcf819e2e395fbe73b56e0a387bd64222e831fd610270cd7ea2505549758bf75c05a994a6d034f65f8f0e6fdcaeab1a34d4a6b4b636e070a38bce737'
+				}
+			];
+			var hmacSHA256Vectors = [
+				{
+					k: '0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20',
+					m: 'cdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcd',
+					h: '372efcf9b40b35c2115b1346903d2ef42fced46f0846e7257bb156d3d7b30d3f'
+				}
+			];
+
+			if (sodium.crypto_auth) for (var i = 0; i < hmacSHA512256Vectors.length; i++) testHmacSHA512256Vector(hmacSHA512256Vectors[i]);
+			if (sodium.crypto_auth_hmacsha512) for (var i = 0; i < hmacSHA512Vectors.length; i++) testHmacSHA512Vector(hmacSHA512Vectors[i]);
+			if (sodium.crypto_auth_hmacsha256) for (var i = 0; i < hmacSHA256Vectors.length; i++) testHmacSHA256Vector(hmacSHA256Vectors[i]);
```

Comment:
Hi @BatikhSouri,

Thank you very much for your detailed explanation. These are very valuable information. Have you been successful building a `libsodium.js` variant with `hmacsha256`?

If not, then I will try to build one because I need that functionality.

Thanks again for your support!

Benny